File: 02_Sorting_Searching/60_Sliding_Window/BaseCode.java
public class BaseCode {

    /**
     * 크기가 K인 연속된 부분 배열 중 가장 큰 합을 반환합니다.
     */
    public static int slidingWindowMaxSum(int[] arr, int k) {
        int n = arr.length;
        
        // 예외 처리
        if (n < k) {
            throw new IllegalArgumentException("배열의 길이가 K보다 작습니다.");
        }

        // 1. 첫 번째 창문(초기 윈도우)의 합 계산
        int windowSum = 0;
        for (int i = 0; i < k; i++) {
            windowSum += arr[i];
        }
        
        int maxSum = windowSum;

        // 2. 창문을 한 칸씩 이동하며 상태 갱신
        for (int i = k; i < n; i++) {
            // 새로 들어온 값(arr[i])은 더하고, 창문 밖으로 밀려난 값(arr[i - k])은 뺌
            windowSum += arr[i] - arr[i - k];
            
            // 최대 합 갱신 (Math.max를 사용하여 간결하게 표현)
            maxSum = Math.max(maxSum, windowSum);
        }

        return maxSum;
    }

    public static void main(String[] args) {
        int[] data = {2, 1, 5, 1, 3, 2};
        int windowSize = 3;

        int result = slidingWindowMaxSum(data, windowSize);
        System.out.println("크기가 " + windowSize + "인 부분 배열의 최대 합: " + result); 
        // Output: 9
    }
}

File: 02_Sorting_Searching/60_Sliding_Window/base_code.py
def sliding_window_max_sum(arr, k):
    """
    크기가 K인 연속된 부분 배열 중 가장 큰 합을 반환하는 슬라이딩 윈도우 템플릿
    """
    n = len(arr)
    
    # 예외 처리: 배열의 길이가 창문 크기보다 작으면 계산 불가
    if n < k:
        return None 
    
    # 1. 첫 번째 창문(인덱스 0부터 k-1까지)의 합을 먼저 구합니다.
    window_sum = sum(arr[:k])
    max_sum = window_sum
    
    # 2. 창문을 한 칸씩 오른쪽으로 밀면서 최대 합을 갱신합니다.
    for i in range(k, n):
        # 핵심 로직: 새로 들어온 값은 더하고, 창문을 벗어난 맨 앞의 값은 뺍니다.
        window_sum += arr[i] - arr[i - k]
        
        if window_sum > max_sum:
            max_sum = window_sum
            
    return max_sum

# 사용 예시
if __name__ == "__main__":
    data = [2, 1, 5, 1, 3, 2]
    window_size = 3
    
    result = sliding_window_max_sum(data, window_size)
    print(f"크기가 {window_size}인 부분 배열의 최대 합: {result}") # Output: 9 (5 + 1 + 3)